### ISSUE TITLE
RocksDB file permissions are not configurable (POSIX)

### ISSUE THREAD

**Issue by ThorAsgardDev** [facebook/rocksdb#11503]
### Expected behavior
Use the permissions of the user running the process that involves RocksDB, or add a function in RocksDB to set the permissions.
I need to have the permissions 664 (or 660).

### Actual behavior
Currently we can choose only between permissions 0644 and 0600.

### Steps to reproduce the behavior
Create a RocksDB database on Linux.

**Comment by jowlyzhang** [comment]
Users of RocksDB usually don't need to directly access the data files RocksDB generated. Can you share a bit more why you need such access?

**Comment by ThorAsgardDev** [comment]
Hello jowlyzhang,
Thanks for your reply.

This is my use case:
I have a program "A" launched by a service using a technical user "techuser". This program uses RocksDB to populate a base.
I have another program "B" launched by a user "humanuser". It is a tool. This program can "modify" an existing RocksDB base created by "A".
"techuser" and "humanuser" are in the same group "databaseadmin".
With this configuration it currently doesn't work because RocksDB does not allow to set the write permissions on group.

**Comment by jowlyzhang** [comment]
You should be able to update the permissions to files generated by RocksDB to any random combination of permissions as long as you update it as an administrator with the proper permissions. It's however risky to modify files generated by RocksDB indirectly (through ways other than RocksDB itself). The files are usually encoded and checksummed, any modification would corrupt the file and corrupt the database. So any permission beyond read-only is not safe.

**Comment by ThorAsgardDev** [comment]
When I say "This program can "modify" an existing RocksDB base created by "A"." I mean "open database using RocksDB to add/remove records".
As the database has been created using the "techuser" so "humanuser" cannot open the database to add/remove records because the database modifications can only be done by "techuser".

**Comment by jowlyzhang** [comment]
I see what you mean now. Two different users launch two different RocksDB processes that access the same set of data files. Regardless of the proper file permission setup to allow this, which should be doable, it's operating system level file permission setup. Please check our wiki page and note that "Only single instance of Primary is allowed" https://github.com/facebook/rocksdb/wiki/Read-only-and-Secondary-instances.  As "add/remove records" are write operations, it requires opening RocksDB as a primary instance.

**Comment by jowlyzhang** [comment]
Have you tried these steps to change the file permission for all the RocksDB files for "humanuser"  (from ChatGPT):

Here's an example of how you can change the permissions for a specific user:

1) Determine if the user is the owner of the file:
    `ls -l filename`
This command will display the file's details, including the owner and group. If the user name matches the owner, you can proceed to step 2. Otherwise, continue to step 3.

2) Use chmod to set permissions for the owner:
   `chmod u+rw filename`
This command grants read and write permissions to the owner.

3) If the user is part of the group associated with the file, use chgrp to change the group ownership:
   `chgrp groupname filename`
Replace groupname with the desired group name.

4) Use chmod to set permissions for the group:
  `chmod g+rw filename`
This command grants read and write permissions to the group.

5) If necessary, set permissions for others using chmod:
   `chmod o+rw filename`

**Comment by mrambacher** [comment]
I understand your issue, and there is no way currently to accomplish this without a change to the code.  believe the same issue might occur if user B attempted to open a database as secondary or even read-only (because of the ability to write the log file).  

Also note that simply changing permissions on files may not work in your case, as the database directories are created with a permissions of 755 -- meaning user B could not write into the database directory either (hence the potential issue with secondaries).

I agree that it would be nice if RocksDB provided this functionality (there is code that attempts to use 644 or 600 that is not exposed -- see #8109).

As a workaround, can you do anything with listeners (implement a listener for when files are created so that you can change their permissions accordingly).  I do not know if this will work, but might be a (long) thread to pull on until a PR is in place.

What version of RocksDB are you using and in what language is your application written?

**Comment by jowlyzhang** [comment]
> I understand your issue, and there is no way currently to accomplish this without a change to the code. believe the same issue might occur if user B attempted to open a database as secondary or even read-only (because of the ability to write the log file).
> 
> Also note that simply changing permissions on files may not work in your case, as the database directories are created with a permissions of 755 -- meaning user B could not write into the database directory either (hence the potential issue with secondaries).
> 
> I agree that it would be nice if RocksDB provided this functionality (there is code that attempts to use 644 or 600 that is not exposed -- see #8109).
> 
> As a workaround, can you do anything with listeners (implement a listener for when files are created so that you can change their permissions accordingly). I do not know if this will work, but might be a (long) thread to pull on until a PR is in place.
> 
> What version of RocksDB are you using and in what language is your application written?

Thanks Mark for the detailed input. I wasn't aware that running an existing RocksDB database as a different user has these hiccups. CC @anand1976

**Comment by ThorAsgardDev** [comment]
Yes that is currently what I have to do but I have to be "root" or "techuser" to change the file properties.
1) "techuser": It is a technical user so a human cannot use this user to run the tool "B".
2) "root": Too many rights, it is not secure.
So currently I don't have any "acceptable" solution.

**Comment by ThorAsgardDev** [comment]
> I understand your issue, and there is no way currently to accomplish this without a change to the code. believe the same issue might occur if user B attempted to open a database as secondary or even read-only (because of the ability to write the log file).
> 
> Also note that simply changing permissions on files may not work in your case, as the database directories are created with a permissions of 755 -- meaning user B could not write into the database directory either (hence the potential issue with secondaries).
> 
> I agree that it would be nice if RocksDB provided this functionality (there is code that attempts to use 644 or 600 that is not exposed -- see #8109).
> 
> As a workaround, can you do anything with listeners (implement a listener for when files are created so that you can change their permissions accordingly). I do not know if this will work, but might be a (long) thread to pull on until a PR is in place.
> 
> What version of RocksDB are you using and in what language is your application written?

Currently I'm using RocksDB 7.8.3 (but I can upgrade to the last version) and my application is written in Java.
